### PR TITLE
CATROID-880 Formula editor buttons for list, text, maths

### PR DESCRIPTION
**Pull request by Piki2000**
https://jira.catrob.at/browse/CATROID-880

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorListsListTest.kt [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+
+@Category(Cat.AppUi::class, Level.Smoke::class)
+@RunWith(Parameterized::class)
+class FormulaEditorListsListTest(
+    private val formulaEditorFunction: String,
+    private val formulaEditorFunctionParameter: String,
+    @Suppress("unused")
+    private val testName: String
```
Comment (juliajulie95) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorListsListTest.kt:57]: Please check why this ios suppressed, could we delete the variable?

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorLogicListTest.kt [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+
+@Category(Cat.AppUi::class, Level.Smoke::class)
+@RunWith(Parameterized::class)
+class FormulaEditorLogicListTest(
+    private val formulaEditorFunction: String,
+    private val formulaEditorFunctionParameter: String,
+    @Suppress("unused")
+    private val testName: String
```
Comment (juliajulie95) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorLogicListTest.kt:53]: Please check why this ios suppressed, could we delete the variable?

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorTextListTest.kt [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+import androidx.test.espresso.action.ViewActions
+import androidx.test.espresso.matcher.ViewMatchers.withId
+
+@Category(AppUi::class, Smoke::class)
+@RunWith(Parameterized::class)
+class FormulaEditorTextListTest(
+    private val formulaEditorFunction: String,
+    private val formulaEditorFunctionParameter: String,
+    @Suppress("unused")
+    private val testName: String
```
Comment (juliajulie95) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorTextListTest.kt:54]: Please check why this is suppressed, could we delete the variable?

**Review by juliajulie95 — CHANGES_REQUESTED** [review]

**Review by juliajulie95 — APPROVED** [review]
Thank you!

**Review by wslany — CHANGES_REQUESTED** [review]
Thanks from me as well. Several things that caught my eye and I would like you to correct:

There are several reporters, sensors etc in the new tabs that are shown there, even though the corresponding AI feature has not been activated, and even though they are then correctly not shown in the other, existing categories. For instance, the OCR or voice recognition related sensors should only be shown in the text category when the OCR and/or voice recognition has been turned on in the settings. Here some corresponding screenshots:
image
image
image
image

This implies that there are also missing tests for the AI extension features in the formula editor, because the incorrect visibility of these sensors in the lists under the AI settings should otherwise have been detected.

On the other hand, the following sensor is not listed under Logic when the corresponding AI extension "Object recognition" has been activated by the user (the screenshot is from the sensor list in the formula editor, where it correctly is shown):
image

Note: Most of the above was not specified in https://jira.catrob.at/browse/CATROID-880 since at the time when the ticket was written, the settings screen for the AI extension had not yet been introduced, and thus this should be changed even though it was not specified in the ticket. Sorry about that!

For the same reasons, there need to be a few other changes, on the one hand in the order the AI sensors are shown in the lists, and on the other hand, subtitles are needed in the "text" and "logic" lists. These subtitles have already been introduced in the sensors category because of the AI extension since the ticket was written, as follows:

When an AI extension is activated, the corresponding sensors shall be listed _above_ other sensors and reporters in the lists (like it is already done in the existing sensors list), together with subtitles above the corresponding sensors. Here how it looks in the sensor list:
image

So, for instance, when the AI sensors in the new "text" list are shown, they should be preceded by corresponding subtitles (the same as in the sensor category), and these subcategories should be shown at the top instead at the bottom, see below:
image

Same order and subtitles in a similar way also for the sensors in the "logic" list. 

Since there must be some delimiter between the special AI extension sensors and the normal text reporters and functions (since the AI sensors shall be shown _above_ the others), please also introduce a new subtitle called "Other text reporters". This new subtitle shall, however, only be shown when some corresponding AI extensions has been activated that needs to be shown in the text list _above_ the normal text reporters, since otherwise there are no subcategories for text. 

Note that no new subtitle for regular reporters is not needed in the logic list, since there is already a subtitle shown anyway because of the different subcategories of regular logic reporters and constants.  

Also note that no subtitle needs to be shown at the top of the "list" functions, since there are no subcategories.

Please also add corresponding tests in a test-first (test driven) way.

**Comment by wslany** [comment]
One more question: Why did you change the comparison color here?
image

**Pull request closed without merging**